src/gen9_avc_encoder.c

```diff
@@ -3596,6 +3662,456 @@ gen9_avc_set_curbe_mbenc(VADriverContextP ctx,
+    }
+
+    avc_state->block_based_skip_enable = cmd->dw3.block_based_skip_enable;
+
+    if (avc_state->rolling_intra_refresh_enable) {
+        //Not supported
+        cmd->dw34.widi_intra_refresh_en = avc_state->rolling_intra_refresh_enable;
+
+    } else {
+        cmd->dw34.widi_intra_refresh_en = 0;
+    }
+    cmd->dw34.enable_per_mb_static_check = avc_state->sfd_enable && generic_state->hme_enabled;
+    cmd->dw34.enable_adaptive_search_window_size = avc_state->adaptive_search_window_enable;
+
+    /* Fixme: Skipped ROI stuffs for now */
+
+    /* r64: FEI specific fields */
+    cmd->dw64.fei_enable = 1;
+    cmd->dw64.multiple_mv_predictor_per_mb_enable = fei_param->mv_predictor_enable;
+    if (fei_param->distortion != VA_INVALID_ID)
+        cmd->dw64.vme_distortion_output_enable = 1;
+    cmd->dw64.per_mb_qp_enable = fei_param->mb_qp;
+    cmd->dw64.mb_input_enable = fei_param->mb_input;
+
+    // FEI mode is disabled when external MVP is available
+    if (fei_param->mv_predictor_enable)
+        cmd->dw64.fei_mode = 0;
+    else
+        cmd->dw64.fei_mode = 1;
+
+    cmd->dw80.mb_data_surf_index = GEN9_AVC_MBENC_MFC_AVC_PAK_OBJ_INDEX;
+    cmd->dw81.mv_data_surf_index = GEN9_AVC_MBENC_IND_MV_DATA_INDEX;
+    cmd->dw82.i_dist_surf_index = GEN9_AVC_MBENC_BRC_DISTORTION_INDEX;
+    cmd->dw83.src_y_surf_index = GEN9_AVC_MBENC_CURR_Y_INDEX;
+    cmd->dw84.mb_specific_data_surf_index = GEN9_AVC_MBENC_MB_SPECIFIC_DATA_INDEX;
+    cmd->dw85.aux_vme_out_surf_index = GEN9_AVC_MBENC_AUX_VME_OUT_INDEX;
+    cmd->dw86.curr_ref_pic_sel_surf_index = GEN9_AVC_MBENC_REFPICSELECT_L0_INDEX;
+    cmd->dw87.hme_mv_pred_fwd_bwd_surf_index = GEN9_AVC_MBENC_MV_DATA_FROM_ME_INDEX;
+    cmd->dw88.hme_dist_surf_index = GEN9_AVC_MBENC_4XME_DISTORTION_INDEX;
+    cmd->dw89.slice_map_surf_index = GEN9_AVC_MBENC_SLICEMAP_DATA_INDEX;
+    cmd->dw90.fwd_frm_mb_data_surf_index = GEN9_AVC_MBENC_FWD_MB_DATA_INDEX;
+    cmd->dw91.fwd_frm_mv_surf_index = GEN9_AVC_MBENC_FWD_MV_DATA_INDEX;
+    cmd->dw92.mb_qp_buffer = GEN9_AVC_MBENC_MBQP_INDEX;
+    cmd->dw93.mb_brc_lut = GEN9_AVC_MBENC_MBBRC_CONST_DATA_INDEX;
+    cmd->dw94.vme_inter_prediction_surf_index = GEN9_AVC_MBENC_VME_INTER_PRED_CURR_PIC_IDX_0_INDEX;
+    cmd->dw95.vme_inter_prediction_mr_surf_index = GEN9_AVC_MBENC_VME_INTER_PRED_CURR_PIC_IDX_1_INDEX;
+    cmd->dw96.mb_stats_surf_index = GEN9_AVC_MBENC_MB_STATS_INDEX;
+    cmd->dw97.mad_surf_index = GEN9_AVC_MBENC_MAD_DATA_INDEX;
+    cmd->dw98.force_non_skip_mb_map_surface = GEN9_AVC_MBENC_FORCE_NONSKIP_MB_MAP_INDEX;
+    cmd->dw99.widi_wa_surf_index = GEN9_AVC_MBENC_WIDI_WA_INDEX;
+    cmd->dw100.brc_curbe_surf_index = GEN9_AVC_MBENC_BRC_CURBE_DATA_INDEX;
+    cmd->dw101.static_detection_cost_table_index = GEN9_AVC_MBENC_SFD_COST_TABLE_INDEX;
+    cmd->dw102.fei_mv_predictor_surf_index = GEN9_AVC_MBENC_MV_PREDICTOR_INDEX;
+    i965_gpe_context_unmap_curbe(gpe_context);
+
+    return;
+}
```

Comment:
The non-fei use case function for mbenc is even more bigger, 500+.
I would prefer to keep a single function since the code inside is not anything complicated, but rather setting values for different attributes.